src/EnergyPlus/PlantLoopHeatPumpEIR.cc

```diff
@@ -552,6 +643,90 @@ void EIRPlantLoopHeatPump::calcSourceSideHeatTransferASHP(EnergyPlusData &state)
+{
+    // energy balance on heat pump
+    this->heatRecoveryRate = this->calcQheatRecovery(this->loadSideHeatTransfer, this->powerUsage);
+    Real64 heatRecoverRateTot = this->heatRecoveryRate;
+
+    // calculate heat recovery side outlet conditions
+    auto &thisHeatRecoveryPlantLoop = state.dataPlnt->PlantLoop(this->heatRecoveryPlantLoc.loopNum);
+    Real64 const CpHR = FluidProperties::GetSpecificHeatGlycol(state,
+                                                               thisHeatRecoveryPlantLoop.FluidName,
+                                                               this->heatRecoveryInletTemp,
+                                                               thisHeatRecoveryPlantLoop.FluidIndex,
+                                                               "EIRPlantLoopHeatPump::calcHeatRecoveryHeatTransferASHP()");
+    Real64 const hRecoveryMCp = this->heatRecoveryMassFlowRate * CpHR;
+    if (this->heatRecoveryMassFlowRate > 0.0) {
+        this->heatRecoveryOutletTemp = this->calcHROutletTemp(this->heatRecoveryInletTemp, this->heatRecoveryRate / hRecoveryMCp);
+    } else {
+        this->heatRecoveryOutletTemp = this->heatRecoveryInletTemp;
+        this->heatRecoveryRate = 0.0;
+    }
+    // limit the HR HW outlet temperature to the maximum allowed (HW Recovery)
+    if (this->heatRecoveryOutletTemp > this->maxHeatRecoveryTempLimit) {
+        if (this->heatRecoveryInletTemp < this->maxHeatRecoveryTempLimit) {
+            this->heatRecoveryOutletTemp = this->maxHeatRecoveryTempLimit;
+            this->heatRecoveryRate = hRecoveryMCp * (this->heatRecoveryOutletTemp - this->heatRecoveryInletTemp);
+        } else {
+            this->heatRecoveryRate = 0.0;
+            this->heatRecoveryOutletTemp = this->heatRecoveryInletTemp;
+        }
+    }
+    // limit the HR CW outlet temp to the minimum allowed (CW Recovery)
+    if (this->heatRecoveryOutletTemp < this->minHeatRecoveryTempLimit) {
+        if (this->heatRecoveryInletTemp > this->minHeatRecoveryTempLimit) {
+            this->heatRecoveryOutletTemp = this->minHeatRecoveryTempLimit;
+            this->heatRecoveryRate = hRecoveryMCp * (this->heatRecoveryInletTemp - this->heatRecoveryOutletTemp);
+        } else {
+            this->heatRecoveryRate = 0.0;
+            this->heatRecoveryOutletTemp = this->heatRecoveryInletTemp;
+        }
+    }
+    // report the net heat balance as source side heat transfer
+    Real64 heatReoveryRateUnused = std::max(0.0, (heatRecoverRateTot - this->heatRecoveryRate));
+    if (heatReoveryRateUnused > 0.0) {
+        this->sourceSideHeatTransfer = heatReoveryRateUnused;
+        // calculate source side outlet conditions
+        Real64 const CpSrc = Psychrometrics::PsyCpAirFnW(state.dataEnvrn->OutHumRat);
+        Real64 const sourceMCp = this->sourceSideMassFlowRate * CpSrc;
+        this->sourceSideOutletTemp = this->calcSourceOutletTemp(this->sourceSideInletTemp, this->sourceSideHeatTransfer / sourceMCp);
```

Comment:
if the heat recovery rate is reduced due to temperature limit, the total or the balance amount is reported as the source-side heat transfer rate and a new source side outlet temperature is determined.